### PR TITLE
MM-52173 Add setting to delay channel autocomplete

### PR DESCRIPTION
**Pull request by hmhealey**
#### Summary
The channel autocomplete can be a bit obtrusive in certain cultures such as Korean where tildes are used more regularly in messages, so we've added a setting to make it require typing two letters after the tilde before the autocomplete triggers.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-52173

#### Related Pull Requests
https://github.com/mattermost/mattermost-server/pull/22952

#### Device Information
This PR was tested on: 

#### Screenshots
https://user-images.githubusercontent.com/3277310/233412110-1ee2e9f3-a469-4e8c-9fa4-3ab466332cd9.mp4

#### Release Note
```release-note
Added support for ExperimentalSettings.DelayChannelAutocomplete
```

### CODE REVIEW

File: app/components/autocomplete/channel_mention/index.ts
Comment (enahum) [reply app/components/autocomplete/channel_mention/index.ts:88]: this is fine, but I would ask if you can remove the extra enhancer and maybe get the `matchTerm` as an observable that you can pipe to get the `localChannels`. My only reason for this is to avoid having `matchPattern` as a prop of the component

Comment (hmhealey) [reply app/components/autocomplete/channel_mention/index.ts:88]: Oh okay, so move this logic to `enhanced` then rather than having it in a separate enhancer? I don't entirely know why I decided I didn't want to access the database in `enhanced` to get the config value there

Comment (hmhealey) [reply app/components/autocomplete/channel_mention/index.ts]: This should be fixed now

**Review by enahum — APPROVED** [review]
LGTM

File: app/components/autocomplete/channel_mention/index.ts
Comment (larkox) [inline app/components/autocomplete/channel_mention/index.ts:84]: This returns an observable, so it should be called `observeX` instead of `getX`.

File: app/components/autocomplete/channel_mention/index.ts
Comment (larkox) [inline app/components/autocomplete/channel_mention/index.ts:100]: 0/5
```suggestion
    const matchTerm = observeMatchPattern(isSearch,  database).pipe(
        map((regexp) => getMatchTermForChannelMention(value.substring(0, cursorPosition), regexp, isSearch))
     );
```

**Review by larkox — APPROVED** [review]
Approving with two minor nits.

File: app/components/autocomplete/channel_mention/index.ts
Comment (hmhealey) [reply app/components/autocomplete/channel_mention/index.ts:100]: Yeah, I wasn't quite happy with how that looked. If this was just chaining methods like `.pipeMap`, then I'd have a way to do it, but the combination of functions is throwing me off a bit.

I was trying the change you suggested, but then I realized I could just do it like some of the other code in the file does it where `pipe` and `map` are on the same line which I like
```js
    const matchTerm = observeMatchPattern(isSearch, database).pipe(map((regexp) => {
        return getMatchTermForChannelMention(value.substring(0, cursorPosition), regexp, isSearch);
    }));
```

**Comment by hmhealey** [comment]
For QA testing this one, it'll need to be pointed at the test server for https://github.com/mattermost/mattermost-server/pull/22952 until that's merged

**Review by stevemudie — APPROVED** [review]
Looks good!

**Pull request merged**
